opencv_ai/splice/component.h

```diff
@@ -0,0 +1,60 @@
+#ifndef OPENCV_AI_SPLICE_COMPONENT_H_
+#define OPENCV_AI_SPLICE_COMPONENT_H_
+
+#include <boost/optional.hpp>
+#include <boost/regex.hpp>
+
+#include <glog/logging.h>
+
+#include <opencv_ai/splice/context.h>
+
+namespace opencv_ai {
+
+inline bool ValidName(const std::string& name) {
+  static const boost::regex e("[a-zA-Z][\\w_]*");
+  return regex_match(name, e);
+}
+
+inline std::string NameJoin(const std::string& parent,
+                            const std::string& child) {
+  return parent + "/" + child;
+}
+
+class Component {
+ public:
+  explicit Component(Context ctx, std::string name,
+                     ContextState state = ContextState::kSame)
+      : ctx_(ctx, state), name_(std::move(name)) {
+    CHECK(ValidName(name_)) << "Name is not valid." << name_;
+  }
+
+  explicit Component(const Component& parent, std::string name,
+                     ContextState state = ContextState::kSame)
+      : ctx_(parent.GetContext(), state),
+        name_(std::move(name)),
+        full_name_(NameJoin(parent.FullName(), name_)),
+        parent_(parent) {
+    CHECK(ValidName(name_)) << "Name is not valid." << name_;
+  }
+
+  Context& GetContext() { return ctx_; }
```

Comment:
Input and Output grab the Context if constructed from a Component.  I guess it could be ``const&`` though.  We can re-evaluate this.  Perhaps it could be a framework access only function.

I'm not really sure what the affects of giving away the context are yet...  Maybe it should be private for now, with a friend access-or.